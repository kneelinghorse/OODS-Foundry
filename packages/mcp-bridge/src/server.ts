import cors from '@fastify/cors';
import rateLimit from '@fastify/rate-limit';
import fastifyStatic from '@fastify/static';
import Fastify, { type FastifyReply, type FastifyRequest } from 'fastify';
import { spawn, ChildProcessWithoutNullStreams } from 'node:child_process';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { bridgeConfig, lowerCaseHeaders } from './config.js';
import { registerArtifactEndpoints } from './endpoints/artifacts.js';
import { buildErrorPayload, normalizeRunErrorCode, sendError, statusForCode } from './middleware/errors.js';

const ALLOWED_TOOLS = new Set(bridgeConfig.tools.allowed);
const WRITE_GATED_TOOLS = new Set(bridgeConfig.tools.writeGated);

type RunRequestBody = {
  tool?: string;
  input?: Record<string, any>;
};

type McpResponse =
  | { id?: string | number; result: any }
  | { id?: string | number; error: { code: string; message?: string; messages?: any } };

class McpClient {
  private child: ChildProcessWithoutNullStreams | null = null;
  private seq = 0;
  private pending = new Map<
    number,
    { resolve: (v: any) => void; reject: (e: any) => void }
  >();
  private buffer = '';

  constructor(private serverCwd: string) {}

  private ensure() {
    if (this.child && !this.child.killed) return;
    // Spawn the built MCP server (dist/index.js)
    const entry = path.join(this.serverCwd, 'dist', 'index.js');
    this.child = spawn(process.execPath, [entry], {
      cwd: this.serverCwd,
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env },
    });

    this.child.stdout.setEncoding('utf8');
    this.child.stdout.on('data', (chunk: string) => {
      this.buffer += chunk;
      let idx: number;
      while ((idx = this.buffer.indexOf('\n')) >= 0) {
        const line = this.buffer.slice(0, idx).trim();
        this.buffer = this.buffer.slice(idx + 1);
        if (!line) continue;
        try {
          const msg = JSON.parse(line) as McpResponse;
          const id = (msg as any).id as number | undefined;
          if (id != null && this.pending.has(id)) {
            const p = this.pending.get(id)!;
            this.pending.delete(id);
            if ((msg as any).error) p.reject((msg as any).error);
            else p.resolve((msg as any).result);
          }
        } catch {
          // ignore parse errors on stdout
        }
      }
    });

    this.child.on('exit', () => {
      // Reject all in-flight requests
      for (const [, p] of this.pending) p.reject({ code: 'PROCESS_EXIT' });
      this.pending.clear();
      this.child = null;
    });
  }

  async run(tool: string, input: any): Promise<any> {
    this.ensure();
    const id = ++this.seq;
    const payload = JSON.stringify({ id, tool, input }) + '\n';
    return new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      this.child!.stdin.write(payload, 'utf8');
    });
  }
}


function isJsonContentType(header: unknown): boolean {
  if (typeof header !== 'string') return false;
  const [mediaType] = header.split(';', 1);
  return mediaType.trim().toLowerCase() === 'application/json';
}

function ensureJsonContentType(request: FastifyRequest, reply: FastifyReply): boolean {
  if (request.method !== 'POST') return true;
  const header = request.headers['content-type'];
  if (isJsonContentType(Array.isArray(header) ? header[0] : header)) return true;
  sendError(reply, 415, 'VALIDATION_ERROR', 'application/json body required.', {
    details: { reason: 'UNSUPPORTED_MEDIA_TYPE' },
  });
  return false;
}

function ensureAuthToken(request: FastifyRequest, reply: FastifyReply): boolean {
  const expected = bridgeConfig.auth.token;
  if (!expected) return true;
  const value = request.headers[lowerCaseHeaders.auth];
  const provided = Array.isArray(value) ? value[0] : value;
  if (!provided) {
    sendError(reply, 401, 'POLICY_DENIED', 'Bridge token required to run tools.', {
      details: { reason: 'MISSING_TOKEN', header: bridgeConfig.auth.header },
    });
    return false;
  }
  if (provided !== expected) {
    sendError(reply, 401, 'POLICY_DENIED', 'Bridge token rejected.', {
      details: { reason: 'INVALID_TOKEN' },
    });
    return false;
  }
  return true;
}

function hasApprovalToken(request: FastifyRequest): boolean {
  const value = request.headers[lowerCaseHeaders.approval];
  if (Array.isArray(value)) return value.some((item) => typeof item === 'string' && item.trim().length > 0);
  return typeof value === 'string' && value.trim().length > 0;
}

async function main() {
  const fastify = Fastify({ logger: false });

  await fastify.register(rateLimit, {
    global: false,
    addHeaders: {
      'x-ratelimit-limit': true,
      'x-ratelimit-remaining': true,
      'x-ratelimit-reset': true,
    },
    errorResponseBuilder: (_request, context) =>
      buildErrorPayload('RATE_LIMITED', 'Too many requests - please slow down.', {
        details: {
          limit: context.max,
          resetInSeconds: Math.ceil(context.ttl / 1000),
        },
      }),
  });

  fastify.addHook('onRoute', (routeOptions) => {
    if (routeOptions.url?.startsWith('/artifacts')) {
      routeOptions.config = {
        ...(routeOptions.config ?? {}),
        rateLimit: bridgeConfig.rateLimit.artifacts,
      };
    }
  });

  await fastify.register(cors, {
    origin: bridgeConfig.cors.origin,
    methods: [...bridgeConfig.cors.methods],
    allowedHeaders: [...bridgeConfig.cors.allowHeaders],
    exposedHeaders: [...bridgeConfig.cors.exposeHeaders],
    credentials: bridgeConfig.cors.credentials,
    maxAge: bridgeConfig.cors.maxAgeSeconds,
  });

  const bridgePort = Number(process.env.MCP_BRIDGE_PORT || 4466);
  // Resolve MCP server cwd relative to this file location (works in dev and dist)
  const resolvedServerDir = fileURLToPath(new URL('../../mcp-server/', import.meta.url));
  const mcpServerCwd = resolvedServerDir;
  const client = new McpClient(mcpServerCwd);
  const artifactsRoot = path.join(mcpServerCwd, 'artifacts');

  await fastify.register(fastifyStatic, {
    root: artifactsRoot,
    prefix: '/artifacts/',
    index: false,
    list: false,
    decorateReply: false,
  });

  await registerArtifactEndpoints(fastify, artifactsRoot, {
    list: bridgeConfig.rateLimit.artifacts,
    detail: bridgeConfig.rateLimit.artifacts,
    files: bridgeConfig.rateLimit.artifacts,
    open: bridgeConfig.rateLimit.artifacts,
  });

  fastify.get('/health', async () => ({ status: 'ok', bridge: 'ready' }));

  fastify.get('/tools', { config: { rateLimit: bridgeConfig.rateLimit.tools } }, async () => ({
    tools: Array.from(ALLOWED_TOOLS),
  }));

  fastify.post<{ Body: RunRequestBody }>('/run', { config: { rateLimit: bridgeConfig.rateLimit.run } }, async (request, reply) => {
    if (!ensureJsonContentType(request, reply)) return;
    if (!ensureAuthToken(request, reply)) return;

    const body = request.body ?? {};
    const tool = body.tool;
    if (!tool || typeof tool !== 'string') {
      sendError(reply, 400, 'VALIDATION_ERROR', 'Tool name is required.', {
        details: { reason: 'MISSING_TOOL' },
      });
      return;
    }
    if (!ALLOWED_TOOLS.has(tool)) {
      sendError(reply, 403, 'POLICY_DENIED', `Tool not allowed: ${tool}`, {
        details: { reason: 'FORBIDDEN_TOOL', tool },
      });
      return;
    }

    const input = { ...(body.input ?? {}) };
    const applyRequested = input.apply === true;

    if (applyRequested && !WRITE_GATED_TOOLS.has(tool)) {
      sendError(reply, 403, 'POLICY_DENIED', `Tool ${tool} cannot apply changes via the bridge.`, {
        details: { reason: 'READ_ONLY_TOOL', tool },
      });
      return;
    }

    const approvalGranted = applyRequested && hasApprovalToken(request);
    if (applyRequested && WRITE_GATED_TOOLS.has(tool) && !approvalGranted) {
      sendError(reply, 403, 'POLICY_DENIED', 'Approval token required to apply changes.', {
        details: { reason: 'READ_ONLY_ENFORCED', tool },
      });
      return;
    }

    if (!approvalGranted) {
      input.apply = false;
    }

    try {
      const result = await client.run(tool, input);
      const normalized = {
        ok: true as const,
        tool,
        artifacts: result?.artifacts ?? [],
        transcriptPath: result?.transcriptPath ?? null,
        bundleIndexPath: result?.bundleIndexPath ?? null,
        diagnosticsPath: result?.diagnosticsPath ?? null,
        preview: result?.preview ?? null,
        artifactsDetail: result?.artifactsDetail ?? null,
      };
      return normalized;
    } catch (err: any) {
      const normalizedCode = normalizeRunErrorCode(err?.code);
      const inferredStatus =
        typeof err?.status === 'number'
          ? err.status
          : typeof err?.statusCode === 'number'
          ? err.statusCode
          : undefined;
      const status = inferredStatus ?? statusForCode(normalizedCode);
      const message = typeof err?.message === 'string' && err.message.length ? err.message : 'Failed to run tool.';
      const details = err?.details ?? err?.messages;
      return sendError(reply, status, normalizedCode, message, {
        details,
        incidentId: typeof err?.incidentId === 'string' ? err.incidentId : undefined,
      });
    }
  });

  async function listenWithFallback() {
    try {
      await fastify.listen({ port: bridgePort, host: '127.0.0.1' });
    } catch (err: any) {
      if (err?.code === 'EADDRINUSE' && !process.env.MCP_BRIDGE_PORT) {
        // If default port is busy and no explicit port set, pick ephemeral
        await fastify.listen({ port: 0, host: '127.0.0.1' });
      } else {
        throw err;
      }
    }
    const addr = fastify.server.address();
    const actualPort = typeof addr === 'object' && addr ? (addr as any).port : bridgePort;
    // eslint-disable-next-line no-console
    console.log(`[mcp-bridge] listening on :${actualPort}`);
  }

  await listenWithFallback();
}

main().catch((err) => {
  // eslint-disable-next-line no-console
  console.error('[mcp-bridge] fatal', err);
  process.exit(1);
});
