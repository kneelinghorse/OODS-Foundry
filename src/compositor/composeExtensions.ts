import { cloneElement, createElement, Fragment, isValidElement } from 'react';
import type { ReactElement, ReactNode } from 'react';
import { REGION_ORDER, type CanonicalRegionID } from '../types/regions.js';
import type { RenderContext } from '../types/render-context.js';
import type { Condition, ViewExtension } from '../types/view-extension.js';
import { normalizeViewExtensions, ViewExtensionError } from '../view/extensions.js';

type TargetBuckets<Data> = Map<string, ViewExtension<Data>[]>;

interface RegionAccumulator<Data> {
  readonly sections: ViewExtension<Data>[];
  readonly actions: ViewExtension<Data>[];
  readonly modifiersByTarget: TargetBuckets<Data>;
  readonly modifiersUnbound: ViewExtension<Data>[];
  readonly wrappersByTarget: TargetBuckets<Data>;
  readonly wrappersUnbound: ViewExtension<Data>[];
}

function createAccumulator<Data>(): RegionAccumulator<Data> {
  return {
    sections: [],
    actions: [],
    modifiersByTarget: new Map(),
    modifiersUnbound: [],
    wrappersByTarget: new Map(),
    wrappersUnbound: [],
  };
}

function normalizeNode(node: unknown): ReactNode {
  if (node === undefined) {
    return null;
  }
  return node as ReactNode;
}

function evaluateCondition<Data>(condition: Condition, context: RenderContext<Data>): boolean {
  const expression = condition.expression?.trim();
  if (!expression) {
    return true;
  }

  // eslint-disable-next-line no-new-func
  const executor = new Function('context', `return (${expression});`);
  const result = executor(context);
  return Boolean(result);
}

function evaluateExtensionConditions<Data>(
  extension: ViewExtension<Data>,
  context: RenderContext<Data>
): boolean {
  const { conditions } = extension;
  if (!conditions || conditions.length === 0) {
    return true;
  }

  for (const condition of conditions) {
    try {
      if (!evaluateCondition(condition, context)) {
        return false;
      }
    } catch (error) {
      const err = error as Error;
      throw new ViewExtensionError(
        `Failed to evaluate condition "${condition.id}" on extension "${extension.id}": ${err.message}`,
        extension.id
      );
    }
  }

  return true;
}

function invokeWrapper(result: unknown, child: ReactNode): ReactNode {
  if (typeof result === 'function') {
    return (result as (children: ReactNode) => ReactNode)(child);
  }

  if (isValidElement(result)) {
    const element = result as ReactElement;
    return cloneElement(element, undefined, child);
  }

  if (result && typeof result === 'object') {
    const maybeWrapper = result as { wrap?: (children: ReactNode) => ReactNode };
    if (typeof maybeWrapper.wrap === 'function') {
      return maybeWrapper.wrap(child);
    }
  }

  return child;
}

function invokeModifier(result: unknown, node: ReactNode): ReactNode {
  if (typeof result === 'function') {
    return (result as (current: ReactNode) => ReactNode)(node);
  }

  if (result && typeof result === 'object') {
    const maybeModifier = result as {
      wrap?: (current: ReactNode) => ReactNode;
      props?: Record<string, unknown>;
    };

    if (typeof maybeModifier.wrap === 'function') {
      return maybeModifier.wrap(node);
    }

    if (isValidElement(node) && maybeModifier.props) {
      const element = node as ReactElement;
      return cloneElement(element, maybeModifier.props);
    }
  }

  return node;
}

function applyWrapperChain<Data>(
  node: ReactNode,
  wrappers: readonly ViewExtension<Data>[],
  context: RenderContext<Data>
): ReactNode {
  return wrappers.reduceRight<ReactNode>((child, wrapper) => {
    const result = wrapper.render(context);
    return invokeWrapper(result, child);
  }, node);
}

function applyModifierChain<Data>(
  node: ReactNode,
  modifiers: readonly ViewExtension<Data>[],
  context: RenderContext<Data>
): ReactNode {
  return modifiers.reduce<ReactNode>((current, modifier) => {
    const result = modifier.render(context);
    return invokeModifier(result, current);
  }, node);
}

function withBoundary<Data>(extension: ViewExtension<Data>, content: ReactNode): ReactNode {
  return createElement(
    'div',
    {
      key: extension.id,
      'data-extension-id': extension.id,
      'data-extension-type': extension.type,
      'data-extension-region': extension.region,
      'data-extension-trait': extension.metadata?.sourceTrait,
      style: { display: 'contents' },
    },
    content
  );
}

function renderRegion<Data>(
  accumulator: RegionAccumulator<Data>,
  context: RenderContext<Data>
): ReactNode | null {
  const nodes: ReactNode[] = [];

  for (const section of accumulator.sections) {
    const modifiers = accumulator.modifiersByTarget.get(section.id) ?? [];
    const wrappers = accumulator.wrappersByTarget.get(section.id) ?? [];

    let node = normalizeNode(section.render(context));
    node = applyModifierChain(node, modifiers, context);
    node = withBoundary(section, node);
    node = applyWrapperChain(node, wrappers, context);
    nodes.push(node);
  }

  for (const action of accumulator.actions) {
    const modifiers = accumulator.modifiersByTarget.get(action.id) ?? [];
    const wrappers = accumulator.wrappersByTarget.get(action.id) ?? [];

    let node = normalizeNode(action.render(context));
    node = applyModifierChain(node, modifiers, context);
    node = withBoundary(action, node);
    node = applyWrapperChain(node, wrappers, context);
    nodes.push(node);
  }

  let combined: ReactNode | null;

  if (nodes.length === 0) {
    combined = null;
  } else if (nodes.length === 1) {
    combined = nodes[0]!;
  } else {
    combined = createElement(Fragment, null, ...nodes);
  }

  combined = applyModifierChain(combined, accumulator.modifiersUnbound, context);
  combined = applyWrapperChain(combined, accumulator.wrappersUnbound, context);

  return combined;
}

export function prepareExtensions<Data>(
  extensions: readonly ViewExtension<Data>[],
  context: RenderContext<Data>
): readonly ViewExtension<Data>[] {
  const normalized = normalizeViewExtensions<Data>(extensions);
  return normalized.filter((extension) => evaluateExtensionConditions(extension, context));
}

export function composeExtensions<Data>(
  extensions: readonly ViewExtension<Data>[],
  context: RenderContext<Data>
): Record<CanonicalRegionID, ReactNode> {
  if (!context) {
    throw new ViewExtensionError('composeExtensions requires a RenderContext to operate.');
  }

  const prepared = prepareExtensions(extensions, context);
  const regions = new Map<CanonicalRegionID, RegionAccumulator<Data>>();

  for (const extension of prepared) {
    let accumulator = regions.get(extension.region);
    if (!accumulator) {
      accumulator = createAccumulator<Data>();
      regions.set(extension.region, accumulator);
    }

    switch (extension.type) {
      case 'section':
        accumulator.sections.push(extension);
        break;
      case 'action':
        accumulator.actions.push(extension);
        break;
      case 'modifier':
        if (extension.targetId) {
          const list = accumulator.modifiersByTarget.get(extension.targetId) ?? [];
          list.push(extension);
          accumulator.modifiersByTarget.set(extension.targetId, list);
        } else {
          accumulator.modifiersUnbound.push(extension);
        }
        break;
      case 'wrapper':
        if (extension.targetId) {
          const list = accumulator.wrappersByTarget.get(extension.targetId) ?? [];
          list.push(extension);
          accumulator.wrappersByTarget.set(extension.targetId, list);
        } else {
          accumulator.wrappersUnbound.push(extension);
        }
        break;
      default:
        break;
    }
  }

  return REGION_ORDER.reduce<Record<CanonicalRegionID, ReactNode>>((result, region) => {
    const accumulator = regions.get(region) ?? createAccumulator<Data>();
    result[region] = renderRegion(accumulator, context);
    return result;
  }, Object.create(null));
}
