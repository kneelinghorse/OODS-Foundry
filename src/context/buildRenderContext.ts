import { createRenderContext } from '../view/context.js';
import type {
  ObjectSpec,
  RenderContext,
  RenderViewport,
  Theme,
} from '../types/render-context.js';

export interface BuildRenderContextInput<Data = unknown> {
  readonly object: ObjectSpec<Data>;
  readonly data: Data;
  readonly theme?: Theme;
  readonly permissions?: readonly string[];
  readonly viewport?: RenderViewport;
}

const DEFAULT_THEME: Theme = Object.freeze({
  id: 'default',
  mode: 'light',
  tokens: Object.freeze({}),
});

const DEFAULT_VIEWPORT: RenderViewport = Object.freeze({
  width: 1280,
  height: 720,
});

export function buildRenderContext<Data = unknown>(
  input: BuildRenderContextInput<Data>
): RenderContext<Data> {
  const {
    object,
    data,
    theme = DEFAULT_THEME,
    permissions = [],
    viewport = DEFAULT_VIEWPORT,
  } = input;

  return createRenderContext<Data>({
    object,
    data,
    theme,
    permissions,
    viewport,
  });
}
