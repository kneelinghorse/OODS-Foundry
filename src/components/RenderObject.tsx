import * as React from 'react';
import type { ReactNode } from 'react';
import { composeExtensions } from '../compositor/composeExtensions.js';
import { buildRenderContext } from '../context/buildRenderContext.js';
import { getContextComponent, type ContextKind } from '../contexts/index.js';
import type { ObjectSpec, RenderContext } from '../types/render-context.js';
import type { RegionMap } from '../types/regions.js';
import type { ViewExtension } from '../types/view-extension.js';
import { resolveTraitExtensions, type TraitViewErrorDetails } from '../traits/adapter.js';
import { renderReport, type RenderReportPayload } from './renderReport.js';

export interface RenderObjectProps<Data = unknown> {
  readonly object: ObjectSpec<Data>;
  readonly context: ContextKind;
  readonly data: Data;
  readonly debug?: boolean;
  readonly className?: string;
}

function useTraitExtensions<Data>(
  object: ObjectSpec<Data>,
  renderContext: RenderContext<Data>,
  onTraitViewError?: (details: TraitViewErrorDetails<Data>) => void
): ViewExtension<Data>[] {
  return React.useMemo(
    () =>
      resolveTraitExtensions<Data>(
        object,
        renderContext,
        onTraitViewError
          ? {
              onTraitViewError,
            }
          : undefined
      ),
    [object, renderContext, onTraitViewError]
  );
}

function useRegionMap<Data>(
  extensions: readonly ViewExtension<Data>[],
  renderContext: RenderContext<Data>
): RegionMap {
  return React.useMemo<RegionMap>(
    () => composeExtensions<Data>(extensions, renderContext) as RegionMap,
    [extensions, renderContext]
  );
}

function useRenderDebug<Data>(
  enabled: boolean | undefined,
  object: ObjectSpec<Data>,
  context: ContextKind,
  regions: RegionMap,
  extensions: readonly ViewExtension<Data>[]
): RenderReportPayload | null {
  const report = React.useMemo<RenderReportPayload | null>(() => {
    if (!enabled) {
      return null;
    }
    return renderReport(object, context, regions, extensions);
  }, [enabled, object, context, regions, extensions]);

  React.useEffect(() => {
    if (!enabled || !report) {
      return;
    }

    const label = `[RenderObject] ${object.name ?? object.id} â†’ ${context}`;
    if (typeof console !== 'undefined') {
      console.groupCollapsed?.(label);
      console.table?.(
        report.regions.map((entry) => ({
          region: entry.region,
          hasContent: entry.hasContent,
          extensions: entry.extensionIds.join(', '),
        }))
      );
      console.info?.('Traits', report.traits);
      console.info?.('Total extensions', report.totalExtensions);
      console.groupEnd?.();
    }
  }, [enabled, report, object, context]);

  return report;
}

export const RenderObject: React.FC<RenderObjectProps> = ({
  object,
  context,
  data,
  debug,
  className,
}) => {
  const renderContext = React.useMemo(
    () =>
      buildRenderContext({
        object,
        data,
      }),
    [object, data]
  );

  const traitErrorLogger = React.useCallback(
    (details: TraitViewErrorDetails<unknown>) => {
      if (!debug) {
        return;
      }

      const objectLabel = object.name ?? object.id;
      const traitLabel = details.trait.id ?? 'unknown-trait';
      const label = `[RenderObject] Trait view failed for "${traitLabel}" on "${objectLabel}" (${context})`;

      if (typeof console !== 'undefined') {
        console.groupCollapsed?.(label);
        console.error?.(details.error);
        console.groupEnd?.();
      }
    },
    [debug, object, context]
  );

  const extensions = useTraitExtensions(
    object,
    renderContext,
    debug ? traitErrorLogger : undefined
  );
  const regions = useRegionMap(extensions, renderContext);
  const report = useRenderDebug(debug, object, context, regions, extensions);

  const ContextComponent = React.useMemo(() => getContextComponent(context), [context]);

  const content = React.useMemo<ReactNode>(() => {
    const renderedContext = <ContextComponent regions={regions} className={className} />;

    if (!debug || !report) {
      return renderedContext;
    }

    return (
      <div
        data-render-debug="true"
        style={{
          display: 'grid',
          gap: '1rem',
          gridTemplateColumns: 'minmax(0, 1fr) minmax(0, 1fr)',
        }}
      >
        <div data-debug-pane="view">{renderedContext}</div>
        <pre data-render-report>
          {JSON.stringify(report, null, 2)}
        </pre>
      </div>
    );
  }, [debug, report, ContextComponent, regions, className]);

  return <>{content}</>;
};

RenderObject.displayName = 'RenderObject';

export type { ContextKind, RenderReportPayload };
