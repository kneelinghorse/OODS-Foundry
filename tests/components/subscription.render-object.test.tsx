import { describe, expect, it } from 'vitest';
import { renderToStaticMarkup } from 'react-dom/server';
import { Fragment, createElement } from 'react';
import type { ReactNode } from 'react';
import { RenderObject } from '../../src/components/RenderObject.js';
import { buildRenderContext } from '../../src/context/buildRenderContext.js';
import { composeExtensions } from '../../src/compositor/composeExtensions.js';
import { resolveTraitExtensions } from '../../src/traits/adapter.js';
import { REGION_ORDER } from '../../src/types/regions.js';
import { createSubscriptionObjectSpec } from '../../src/objects/subscription/object.js';
import type { SubscriptionRecord } from '../../src/objects/subscription/types.js';
import activeSubscription from '../../src/fixtures/subscription/active.json';
import delinquentSubscription from '../../src/fixtures/subscription/past_due.json';
import cancelSubscription from '../../src/fixtures/subscription/active_cancel_at_period_end.json';
import type { ContextKind } from '../../src/contexts/index.js';
import { collectContributionExtensions } from '../../src/engine/contributions/index.js';

const Active = activeSubscription as SubscriptionRecord;
const Delinquent = delinquentSubscription as SubscriptionRecord;
const CancelAtPeriodEnd = cancelSubscription as SubscriptionRecord;

const SubscriptionObject = createSubscriptionObjectSpec();

function renderRegionContent(content: ReactNode): string {
  if (content === undefined || content === null || content === false) {
    return '';
  }

  return renderToStaticMarkup(createElement(Fragment, null, content));
}

function renderContextMarkup(
  data: SubscriptionRecord,
  view: ContextKind = 'detail'
): Record<string, string> {
  const context = buildRenderContext<SubscriptionRecord>({
    object: SubscriptionObject,
    data,
  });
  const baseExtensions = resolveTraitExtensions(SubscriptionObject, context);
  const contributionExtensions = collectContributionExtensions<SubscriptionRecord>({
    object: SubscriptionObject,
    context: view,
    renderContext: context,
  });
  const regions = composeExtensions(
    [...baseExtensions, ...contributionExtensions],
    context
  );

  return REGION_ORDER.reduce<Record<string, string>>((result, region) => {
    result[region] = renderRegionContent(regions[region]);
    return result;
  }, {});
}

describe('Subscription view integration', () => {
  it('renders across contexts without throwing', () => {
    const markup = renderToStaticMarkup(
      createElement(RenderObject, {
        object: SubscriptionObject,
        context: 'detail',
        data: Active,
      })
    );

    expect(markup).toContain('data-view-context="detail"');
  });

  it('renders expected status badges per fixture', () => {
    const activeMarkup = renderContextMarkup(Active);
    const delinquentMarkup = renderContextMarkup(Delinquent);
    const cancelMarkup = renderContextMarkup(CancelAtPeriodEnd);

    expect(activeMarkup.pageHeader).toContain('data-status="active"');
    expect(activeMarkup.pageHeader).toContain('Active');
    expect(delinquentMarkup.pageHeader).toContain('data-status="delinquent"');
    expect(delinquentMarkup.pageHeader).toContain('Delinquent');
    expect(cancelMarkup.pageHeader).toContain('data-status="pending_cancellation"');
    expect(cancelMarkup.pageHeader).toContain('Pending Cancellation');
  });

  it('shows update payment action only for delinquent subscriptions', () => {
    const activeMarkup = renderContextMarkup(Active);
    const delinquentMarkup = renderContextMarkup(Delinquent);
    const cancelMarkup = renderContextMarkup(CancelAtPeriodEnd);

    expect(delinquentMarkup.viewToolbar).toContain('Update payment');
    expect(activeMarkup.viewToolbar).not.toContain('Update payment');
    expect(cancelMarkup.viewToolbar).not.toContain('Update payment');

  });

  it('renders cancellation banner only when scheduled at period end', () => {
    const activeMarkup = renderContextMarkup(Active);
    const cancelMarkup = renderContextMarkup(CancelAtPeriodEnd);

    expect(cancelMarkup.main).toContain('Cancellation scheduled at period end');
    expect(cancelMarkup.main).toMatchSnapshot('main-cancel-at-period-end');
    expect(activeMarkup.main).not.toContain('Cancellation scheduled at period end');
  });

  it('includes renewal summary and progress indicator', () => {
    const activeMarkup = renderContextMarkup(Active);

    expect(activeMarkup.main).toContain('Renews on');
    expect(activeMarkup.main).toContain('Billing');
    expect(activeMarkup.main).toMatchSnapshot('main-renewal-summary');
  });
});
