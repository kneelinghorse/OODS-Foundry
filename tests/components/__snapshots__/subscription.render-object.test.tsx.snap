// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Subscription view integration > includes renewal summary and progress indicator > main-renewal-summary 1`] = `"<div data-extension-id="billable:price-summary" data-extension-type="section" data-extension-region="main" data-extension-trait="Billable" style="display:contents"><div class="rounded-xl border border-slate-200 bg-white p-6 shadow-sm transition-shadow dark:border-slate-800 dark:bg-slate-950 flex flex-col gap-3"><div><h2 class="text-lg leading-7 font-semibold">Billing</h2><p class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">$99.00 / month</p></div><dl class="grid grid-cols-1 gap-3 sm:grid-cols-2"><div class="flex flex-col gap-1"><dt class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">Last payment</dt><dd class="text-base leading-6 font-medium">May 1, 2024</dd></div><div class="flex flex-col gap-1"><dt class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">Next payment due</dt><dd class="text-base leading-6 font-medium">May 31, 2024</dd></div></dl></div></div><div data-extension-id="timestampable:renewal-summary" data-extension-type="section" data-extension-region="main" data-extension-trait="Timestampable" style="display:contents"><div class="rounded-xl border border-slate-200 bg-white p-6 shadow-sm transition-shadow dark:border-slate-800 dark:bg-slate-950 flex flex-col gap-4"><div class="flex flex-col gap-1"><h2 class="text-lg leading-7 font-semibold">Renewal</h2><p class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">Renews on May 31, 2024</p></div><div class="flex flex-col gap-3"><div class="flex items-center justify-between text-xs uppercase tracking-wide text-slate-500 dark:text-slate-400"><span>May 1, 2024</span><span>May 31, 2024</span></div><div class="h-2 overflow-hidden rounded-full bg-slate-200 dark:bg-slate-800"><div class="h-full rounded-full bg-emerald-500 dark:bg-emerald-400" style="width:35%" aria-label="Billing cycle 35% complete"></div></div><p class="text-sm leading-5 font-normal text-right text-slate-500 dark:text-slate-400">35% of cycle elapsed</p></div></div></div><div data-extension-id="cancellable:scheduled-banner" data-extension-type="section" data-extension-region="main" data-extension-trait="Cancellable" style="display:contents"></div>"`;

exports[`Subscription view integration > renders cancellation banner only when scheduled at period end > main-cancel-at-period-end 1`] = `"<div data-extension-id="billable:price-summary" data-extension-type="section" data-extension-region="main" data-extension-trait="Billable" style="display:contents"><div class="rounded-xl border border-slate-200 bg-white p-6 shadow-sm transition-shadow dark:border-slate-800 dark:bg-slate-950 flex flex-col gap-3"><div><h2 class="text-lg leading-7 font-semibold">Billing</h2><p class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">$129.00 / month</p></div><dl class="grid grid-cols-1 gap-3 sm:grid-cols-2"><div class="flex flex-col gap-1"><dt class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">Last payment</dt><dd class="text-base leading-6 font-medium">May 10, 2024</dd></div><div class="flex flex-col gap-1"><dt class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">Next payment due</dt><dd class="text-base leading-6 font-medium">Jun 9, 2024</dd></div></dl></div></div><div data-extension-id="timestampable:renewal-summary" data-extension-type="section" data-extension-region="main" data-extension-trait="Timestampable" style="display:contents"><div class="rounded-xl border border-slate-200 bg-white p-6 shadow-sm transition-shadow dark:border-slate-800 dark:bg-slate-950 flex flex-col gap-4"><div class="flex flex-col gap-1"><h2 class="text-lg leading-7 font-semibold">Renewal</h2><p class="text-sm leading-5 font-normal text-slate-500 dark:text-slate-400">Renews on Jun 9, 2024</p></div><div class="flex flex-col gap-3"><div class="flex items-center justify-between text-xs uppercase tracking-wide text-slate-500 dark:text-slate-400"><span>May 10, 2024</span><span>Jun 9, 2024</span></div><div class="h-2 overflow-hidden rounded-full bg-slate-200 dark:bg-slate-800"><div class="h-full rounded-full bg-emerald-500 dark:bg-emerald-400" style="width:52%" aria-label="Billing cycle 52% complete"></div></div><p class="text-sm leading-5 font-normal text-right text-slate-500 dark:text-slate-400">52% of cycle elapsed</p></div></div></div><div data-extension-id="cancellable:scheduled-banner" data-extension-type="section" data-extension-region="main" data-extension-trait="Cancellable" style="display:contents"><div class="rounded-xl border border-slate-200 bg-white p-6 shadow-sm transition-shadow dark:border-slate-800 dark:bg-slate-950 border-amber-200/70 bg-amber-50/80 text-amber-900 dark:border-amber-500/40 dark:bg-amber-900/30 dark:text-amber-100"><h2 class="text-base leading-6 font-semibold">Cancellation scheduled at period end</h2><p class="text-sm leading-5 font-normal text-amber-800/90 dark:text-amber-200">Reason: Switching to annual billing with reseller</p><p class="text-sm leading-5 font-normal text-amber-700/80 dark:text-amber-300">Requested May 15, 2024</p></div></div>"`;
